game/minecraft_nbt.ksy

```diff
@@ -0,0 +1,225 @@
+
+  ---
+
+  Sample files:
+
+    * <https://wiki.vg/NBT#Download>
+    * <https://github.com/twoolie/NBT/blob/f9e892e/tests/world_test/data/scoreboard.dat>
+    * <https://github.com/chmod222/cNBT/tree/3f74b69/testdata>
+    * <https://github.com/PistonDevelopers/hematite_nbt/tree/0b85f89/tests>
+doc-ref:
+  - https://wiki.vg/NBT
+  - https://web.archive.org/web/20110723210920/https://www.minecraft.net/docs/NBT.txt
+  - https://minecraft.gamepedia.com/NBT_format
+seq:
+  - size: 0
+    if: root_type == tag::end # always `false`, but it doesn't matter
+    doc: Invoke `valid` check before we start parsing
+  - id: root
+    type: named_tag
+instances:
+  root_type:
+    pos: 0
+    type: u1
+    enum: tag
+    valid: tag::compound
+types:
+  named_tag:
+    -webide-representation: 'TAG_{type}("{name.data}"): {payload:dec}'
+    seq:
+      - id: type
+        type: u1
+        enum: tag
+      - id: name
+        type: tag_string
+        if: not is_tag_end
+      - id: payload
+        type:
+          switch-on: type
+          cases:
+            tag::byte: s1
+            tag::short: s2
+            tag::int: s4
+            tag::long: s8
+            tag::float: f4
+            tag::double: f8
+            tag::byte_array: tag_byte_array
+            tag::string: tag_string
+            tag::list: tag_list
+            tag::compound: tag_compound
+            tag::int_array: tag_int_array
+            tag::long_array: tag_long_array
+    instances:
+      is_tag_end:
+        value: type == tag::end
+  tag_byte_array:
+    -webide-representation: '{len_data:dec} bytes'
+    seq:
+      - id: len_data
+        type: s4
+      - id: data
+        size: len_data
+  tag_string:
+    -webide-representation: '{data}'
+    seq:
+      - id: len_data
+        type: u2
+        doc: unsigned according to https://wiki.vg/NBT#Specification
+      - id: data
+        size: len_data
+        type: str
+  tag_list:
+    -webide-representation: '{num_tags:dec} entries of type TAG_{tags_type}'
+    seq:
+      - id: tags_type
+        type: u1
+        enum: tag
+      - id: num_tags
+        type: s4
+      - id: tags
+        type:
+          switch-on: tags_type
+          cases:
+            tag::byte: s1
+            tag::short: s2
+            tag::int: s4
+            tag::long: s8
+            tag::float: f4
+            tag::double: f8
+            tag::byte_array: tag_byte_array
+            tag::string: tag_string
+            tag::list: tag_list
+            tag::compound: tag_compound
+            tag::int_array: tag_int_array
+            tag::long_array: tag_long_array
+        repeat: expr
+        repeat-expr: num_tags
+  tag_compound:
+    -webide-representation: '{dump_num_tags:dec} entries'
+    seq:
+      - id: tags
+        type: named_tag
+        repeat: until
+        repeat-until: _.is_tag_end
+    instances:
+      dump_num_tags:
+        value: 'tags.size - (tags.last.is_tag_end ? 1 : 0)'
```

Comment:
> Is it even allowed to have a compound that doesn't end with an end tag?

No, I'm pretty sure it isn't. There is no way to achieve that: `repeat-until: _.is_tag_end` is correct, meaning that encountering a `TAG_End` is the only valid way to stop parsing `tags`.

The [spec](https://web.archive.org/web/20110723210920/https://www.minecraft.net/docs/NBT.txt) makes it pretty clear:

```
    TYPE: 10 NAME: TAG_Compound
    Payload: A sequential list of Named Tags. This array keeps going until a TAG_End is found.
             TAG_End end
```

The `(tags.last.is_tag_end ? 1 : 0)` thing is just a kind of defensive programming: after spending once several hours at work trying to figure out that some badly-written Wordpress plugin was trying to strip a `https://...` prefix from the beginning of a `http://...` URL using `substring` (but the code didn't check whether the URL really begins with that prefix and it didn't, resulting in that the `substring` ate up 1 extra character), I realized that I don't ever want to make bugs like that. If the assumptions are always correct, then great, it works. But doing a check just to be sure never harms anything (well, probably takes around 1-4 ticks of CPU time) and it's quicker to add it than examine all possible edge cases to find some 100% working assumptions that _might_ lead to code simplification. Moreover, only that an assumption is true for the current state of code doesn't mean that it will remain 100% working in the future when the code is changed.

> Even if a missing end tag is allowed, this check won't work if the compound has no contents at all - it would try to access the last element of an empty array.

Good point, but that does not apply here I believe. The `tags` array cannot end up empty after proper parsing, because it is delimited by `repeat-until` which is a 1+ repetition (the `do { ... } while (...)` statement is usually used; it unconditionally parses the 1st element and then decides whether to parse the 2nd or not).

But yeah, true, I probably should take this case into account too - when we get serialization, it _will_ be possible to make any array empty, and then it could fail.

Reply:
And I believe there is a yet another advantage of `value: 'tags.size - (tags.last.is_tag_end ? 1 : 0)'` over `value: tags.size - 1`: the former makes it explicit what we are removing by doing `- 1`.

Reply:
> But yeah, true, I probably should take this case into account too - when we get serialization, it _will_ be possible to make any array empty, and then it could fail.

Solved in https://github.com/kaitai-io/kaitai_struct_formats/pull/404/commits/1bb30188421aab44d49fe2d2934612f7a8c72630.

@dgelessus Are you OK with this approach or do you find these checks confusing?

Reply:
Now that I know the reason behind it, it makes sense 🙂 Normally I would put this sort of check in an `assert` (to make it clear that it should really always be true), but that doesn't exist in KS, so this is a reasonable workaround.